feat(extensions): auto registration and repo subpaths

avidal opened the pull request:
note: this is a ground-up rework of a previous PR for nested extensions: #6363

also note: the names `prefix` and `path` may be too generic and I'm happy to change them

This patch introduces support for specifying a `prefix` and a `path` when registering an extension repository, each of which influences extension loading.

### prefix

If an extension repo has a `prefix` specified, then extension discovery will attempt to locate an extension repository with this prefix if the extension is not otherwise already registered.

For example:

```
v1alpha1.extension_repo(name="custom", prefix="internal")

load("ext://internal/abc", "...")
```

In this case, the extension loading mechanism will first look for an already registered extension by that name, and if not found, will look for an extension repository with a prefix of `internal`, and then proceed to register the extension for that repo.

The roughly equivalent `v1alpha1.extension` declaration would be: `v1alpha1.extension(name="internal_abc", repo="custom", ...)`

### path

In addition to a repo-wide auto registration prefix this patch introduces a `path` argument for extension repositories. This acts similarly to the `repo_path` argument for an extension, except it applies to *all* extensions located in that repository.

For example, combining prefix and path:

```
v1alpha1.extension_repo(name="custom", url="...", prefix="internal", path="extensions")
v1alpha1.extension(name="custom-root", repo_name="custom")
v1alpha1.extension(name="explicit", repo_name="custom", repo_path="registered")

# Loads from custom:extensions/Tiltfile since the extension has no `repo_path`
load("ext://custom-root", "...")

# Loads from custom:extensions/registered
load("ext://explicit", "...")

# Loads from custom:extensions/abc
load("ext://internal/abc", "...")

# Loads from custom:extensions/abc/def
load("ext://internal/abc/def", "...")
```

## Why?

At Datadog, we have a large Tilt extension library and most of our codebase is in a monorepo. The current extension loading mechanism makes it difficult to organize our extensions in such a way that's both maintainable by the extension authors and not onerous for the users.

These two changes provide a lot more flexibility for managing extension libraries. Using a repo wide `path` allows you to keep your extensions out of the repository root, and using a `prefix` allows your users to just drop a single `extension_repo` at the top of their `Tiltfile`.

With this, our complex extension meta loading extension can be reduced to a single `v1alpha1.extension_repo` line, and our users can `load("ext://tiltlib/...")`

avidal commented:
For reviewers: note that each commit is logically separate:

- A test for the existing "nesting" behavior against the default repository plus a side-effect test (a registered extension named `one_two` can be loaded via `ext://one/two`)
- The repo prefix implementation
- The repo path implementation
- And a few more tests

avidal commented:
Also: if `repo.path` is too far (or too complex) I think that *just* having a repo prefix would mostly solve our needs and I can back out that part.

nicks commented on pkg/apis/core/v1alpha1/extensionrepo_types.go:
```diff
@@ -68,11 +68,24 @@ type ExtensionRepoSpec struct {
+	// For example, an extension repo with prefix="internal" will cause any load()
+	// calls that start with `ext://internal` to be loaded from this repository.
+	// +optional
+	Prefix string `json:"prefix,omitempty" protobuf:"bytes,3,opt,name=prefix"`
```
how do you feel about naming these as:

LoadHost -- used for `load("ext://{LoadHost}/path/to/extension")`
GitSubPath -- subpath in a git repo to find the extension, only valid when URL points to a git repo

i 100% agree that naming and ergonomics is the hardest part!!

avidal replied:
re: GitSubPath

`GitSubPath` works in the general case where the extension repo is an actual git repository, but feels kind of misnamed when you're using a local extension repo. Although I guess in that case you wouldn't need the extra path specifier either (since you can just specify it in the repo path itself).

Maybe there's an opportunity to extend the "url" part to carry a subpath as well? Naively you can do `<host>/<owner>/<repo>/<path>` but that fails for hosts like gitlab or bitbucket where there's often an extra path component already for a "project". To handle that the resolver *could* have special casing for the most common code hosts with a fallback that attempts cloning starting from the full url, chopping off a path segment on each fetch failure until it gets to the base case of `<host>/<repo>` but that seems maybe too convoluted.

I also feel like `extensions_path` matches more closely with `repo_path` in the `extension` API and might be more clear of a name. I think `subpath` gets most of the way there but mostly for whoever is writing the `extension_repo` line in a `Tiltfile`, it may not be clear enough for the other people on the team that are just *reading* the line.

re: LoadHost

I'm not sure what `Host` is meant to indicate. Now that I'm thinking about it more, this could theoretically be simplified a little bit if, instead of a string prefix, this is a boolean that defaults to false, and if true, registers the name of the repository as the "load prefix".

So, something like `extension_repo(name='datadog', url='...', register_extensions=True)` would trigger auto registration for any otherwise unregistered extension that starts with `ext://datadog/`.

nicks replied:
re: what is Host meant to indicate - https://datatracker.ietf.org/doc/html/rfc1738#section-3.1, more as a convention, because it's that part of the URI

re: ya, i think we'd want it to be an error if you use GitSubPath and a file URL.

nicks replied:
and ya, i'd rather avoid sticking more special cases in the URL and lean towards explicit fields

avidal replied:
> re: what is Host meant to indicate - https://datatracker.ietf.org/doc/html/rfc1738#section-3.1, more as a convention, because it's that part of the URI

ohhh, i wasn't even thinking about it as "hostname".

Ultimately I'm happy to change it to whatever you prefer. Do you want `GitSubPath / git_subpath` and `LoadHost / load_host`?

nicks replied:
yes, thank you!

avidal replied:
done in the latest commit(s)

nicks commented:
(not sure what's going on with the compose v1 tests, lemme handle that in a separate pr)

nicks commented:
ya, if you rebase, the Compose error should go away

avidal commented on pkg/apis/core/v1alpha1/extensionrepo_types.go:
```diff
@@ -68,11 +68,24 @@ type ExtensionRepoSpec struct {
+	// A path within the repository root that contains all extensions. By default, all extensions
+	// are expected to live at the repository root.
+	// +optional
+	Path string `json:"path,omitempty" protobuf:"bytes,4,opt,name=path"`
```
changed to `git_subpath`

avidal commented:
@nicks latest version is rebased and has a few more bits and bobs:

- Changed `prefix` to `load_host`
- Changed `path` to `git_subpath`
- Return an error during repo registration for `file://` repos that also have a `git_subpath`
- Split up `p.ensureExtension` into several helper methods (register and register default) with their own tests

There's a few things I'd like to get your opinion on that I'll call out on specific file comments.

avidal commented on internal/tiltfile/tiltextension/plugin.go:
```diff
@@ -129,6 +131,75 @@ func (e *Plugin) LocalPath(t *starlark.Thread, arg string) (localPath string, er
+// In cases where an extension is not already registered, this function will search for an extension
+// repo that can satisfy the requested extension, with a fallback to an extension in the default
+// repository.
+func (e *Plugin) registerExtension(t *starlark.Thread, extSet, repoSet apiset.TypedObjectSet, extName, moduleName string) *v1alpha1.Extension {
```
I vacillated quite a bit on what the sig should be for `registerExtension` and `registerDefaultExtension`. I wasn't sure if they should share the signature with `ensureExtension` and only take `moduleName` and `objSet` and reconstruct `extName` and the specific typed sets, or if they should take the already "prepared" variables from `.ensureExtension`.

Take a look @nicks, particularly at how these signatures manifest in the tests if you would.

avidal commented on internal/tiltfile/tiltextension/plugin.go:
```diff
@@ -129,6 +131,75 @@ func (e *Plugin) LocalPath(t *starlark.Thread, arg string) (localPath string, er
+	existing := extSet[extName]
+	if existing != nil {
+		return existing.(*v1alpha1.Extension)
+	}
```
This seems a bit off. In a fuller implementation this would also set the `tiltfile.loader` annotation but then it's effectively incorporating `ensureExtension`. Maybe there's a better way to break up `ensureExtension`. I'll think on it.

avidal commented on internal/tiltfile/tiltextension/plugin_test.go:
```diff
@@ -191,6 +195,311 @@ include('Tiltfile.b')
+		t.Skip()
+	}
+
+	t.Skip("cannot use file:// repos with git_subpath, making this test unusable")
```
Due to the validation when registering an extension repository we cannot actually run this test after the change to block `file://` repos with a `git_subpath` test.

I don't know if there's a simple way to support `https://` repos in the test without requiring the network.

avidal commented on internal/tiltfile/tiltextension/plugin_test.go:
```diff
@@ -191,6 +195,311 @@ include('Tiltfile.b')
+	extName := apis.SanitizeName(moduleName)
+	extSet := objSet.GetOrCreateTypedSet(&v1alpha1.Extension{})
+
+	ext := p.registerDefaultExtension(nil /* *starlark.Thread */, extSet, extName, moduleName)
```
All of these tests of the extension loading helpers are a bit crufty since I'm not sure how else to get a handle on a `model`, which I need for an `objSet`. I'm also not sure how to get a `*starlark.Thread`, but these functions don't actually *use* the thread so I pass in `nil`.

avidal commented:
It appears the codegen check is failing in CI because of a field name mismatch (the triggering rule is [`names_match`](https://github.com/kubernetes/kube-openapi/blob/master/pkg/generators/rules/names_match.go#L54-L76) seemingly due to using `LoadHost` and `load_host` instead of `LoadHost` and `loadHost`.

Curiously, the `ExtensionSpec` uses `RepoPath` with a json and protobuf key of `repoPath` (to fit the `name_match` rule), but somehow the generated Starlark uses `repo_path`.

nicks reviewed (approved):
LGTM

this seems good! ya, there's some oddities in the test but they don't bother me too much, i can go back and do a pass on making them simpler
